File: apps/ios/GalaxySSI/AgentExecutionPolicy.swift
import Foundation

enum AgentExecutionTaskKind: String, Codable, CaseIterable, Identifiable {
  case chat = "CHAT"
  case research = "RESEARCH"
  case artifact = "ARTIFACT"
  case build = "BUILD"
  case install = "INSTALL"
  case device = "DEVICE"

  var id: String { rawValue }
}

enum AgentExecutionReasoningEffort: String, Codable, CaseIterable, Identifiable {
  case low = "LOW"
  case medium = "MEDIUM"
  case high = "HIGH"

  var id: String { rawValue }
}

struct AgentExecutionProfile: Codable, Equatable {
  var taskKind: AgentExecutionTaskKind
  var reasoningEffort: AgentExecutionReasoningEffort
  var noProgressTimeoutMillis: Int64
  var maxSameFailureAttempts: Int
  var requiresArtifact: Bool
  var targetPlatform: String
  var verifyInstallation: Bool
  var taskIntent: AgentTaskIntent
  var taskIntentConfidence: Int
  var taskIntentSignals: [String]

  init(
    taskKind: AgentExecutionTaskKind,
    reasoningEffort: AgentExecutionReasoningEffort,
    noProgressTimeoutMillis: Int64,
    maxSameFailureAttempts: Int = 2,
    requiresArtifact: Bool = false,
    targetPlatform: String = "",
    verifyInstallation: Bool = false,
    taskIntent: AgentTaskIntent = .chat,
    taskIntentConfidence: Int = 100,
    taskIntentSignals: [String] = []
  ) {
    self.taskKind = taskKind
    self.reasoningEffort = reasoningEffort
    self.noProgressTimeoutMillis = noProgressTimeoutMillis
    self.maxSameFailureAttempts = maxSameFailureAttempts
    self.requiresArtifact = requiresArtifact
    self.targetPlatform = targetPlatform
    self.verifyInstallation = verifyInstallation
    self.taskIntent = taskIntent
    self.taskIntentConfidence = taskIntentConfidence
    self.taskIntentSignals = taskIntentSignals
  }

  enum CodingKeys: String, CodingKey {
    case taskKind = "task_kind"
    case reasoningEffort = "reasoning_effort"
    case noProgressTimeoutMillis = "no_progress_timeout_millis"
    case maxSameFailureAttempts = "max_same_failure_attempts"
    case requiresArtifact = "requires_artifact"
    case targetPlatform = "target_platform"
    case verifyInstallation = "verify_installation"
    case taskIntent = "task_intent"
    case taskIntentConfidence = "task_intent_confidence"
    case taskIntentSignals = "task_intent_signals"
  }

  static func forGoal(
    _ goal: String,
    hasAttachments: Bool = false
  ) -> AgentExecutionProfile {
    let normalized = goal
      .lowercased()
      .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)
    let install = normalized.containsAny(installTerms)
    let build = normalized.containsAny(buildTerms)
    let artifactRequest = normalized.containsAny(artifactTerms)
    let research = normalized.containsAny(researchTerms)
    let intent = AgentTaskIntentClassifier.classify(
      goal: normalized,
      hasAttachments: hasAttachments
    )
    let device = intent.intent == .phoneControl &&
      intent.matchedSignals.contains("phone-control-action")
    let taskKind: AgentExecutionTaskKind
    if install {
      taskKind = .install
    } else if build {
      taskKind = .build
    } else if artifactRequest || hasAttachments {
      taskKind = .artifact
    } else if research {
      taskKind = .research
    } else if device {
      taskKind = .device
    } else {
      taskKind = .chat
    }
    let complex: Set<AgentExecutionTaskKind> = [.research, .artifact, .build, .install]
    let timeoutMillis: Int64
    switch taskKind {
    case .chat:
      timeoutMillis = 180_000
    case .device:
      timeoutMillis = 120_000
    case .research:
      timeoutMillis = 300_000
    case .artifact:
      timeoutMillis = 360_000
    case .build, .install:
      timeoutMillis = 420_000
    }
    return AgentExecutionProfile(
      taskKind: taskKind,
      reasoningEffort: complex.contains(taskKind) ? .medium : .low,
      noProgressTimeoutMillis: timeoutMillis,
      requiresArtifact: artifactRequest || [.build, .install].contains(taskKind),
      targetPlatform: normalized.containsAny(androidTerms) ? "android" : "",
      verifyInstallation: taskKind == .install,
      taskIntent: intent.intent,
      taskIntentConfidence: intent.confidence,
      taskIntentSignals: intent.matchedSignals
    )
  }

  var contract: String {
    var result = "GalaxySSI execution contract: task=\(taskKind.rawValue.lowercased()), intent=\(taskIntent.rawValue.lowercased()), reasoning_effort=\(reasoningEffort.rawValue.lowercased()). Use Plan -> Act -> Observe -> Replan -> Verify -> Finalize. "
    result += "Checkpoint useful work before long or risky actions. "
    result += "Do not repeat an unchanged failing approach. "
    if requiresArtifact {
      result += "A single deliverable remains in its native format; package a directory or multi-file project as ZIP. "
    }
    if verifyInstallation {
      result += "Only report installation or launch after the target runtime returns a verified execution receipt. "
    }
    result += "Do not report success without verification evidence."
    return result
  }

  private static let buildTerms = [
    "build", "compile", "implement", "develop", "write a program", "create an app",
    "create a game", "fix bug", "run tests",
    "\u{7f16}\u{8bd1}", "\u{6784}\u{5efa}", "\u{5f00}\u{53d1}", "\u{5b9e}\u{73b0}",
    "\u{5199}\u{4e00}\u{4e2a}\u{7a0b}\u{5e8f}", "\u{505a}\u{4e00}\u{4e2a}\u{6e38}\u{620f}",
    "\u{751f}\u{6210}\u{7a0b}\u{5e8f}", "\u{4fee}\u{590d} bug", "\u{8fd0}\u{884c}\u{6d4b}\u{8bd5}"
  ]
  private static let installTerms = [
    "install", "install and open", "install apk", "deploy to phone", "launch the app",
    "\u{5b89}\u{88c5}", "\u{5b89}\u{88c5}\u{5e76}\u{6253}\u{5f00}", "\u{5b89}\u{88c5} apk",
    "\u{5b89}\u{88c5}\u{5230}\u{624b}\u{673a}", "\u{7f16}\u{8bd1}\u{5e76}\u{5b89}\u{88c5}"
  ]
  private static let artifactTerms = [
    "return the file", "send the file", "export", "generate image", "create file",
    "downloadable", "zip project", "apk",
    "\u{53d1}\u{56de}\u{6587}\u{4ef6}", "\u{8fd4}\u{56de}\u{6587}\u{4ef6}", "\u{5bfc}\u{51fa}",
    "\u{751f}\u{6210}\u{56fe}\u{7247}", "\u{6253}\u{5305}", "\u{538b}\u{7f29}\u{5305}"
  ]
  private static let researchTerms = [
    "latest", "today", "news", "weather", "research", "search the web",
    "\u{6700}\u{65b0}", "\u{4eca}\u{5929}", "\u{65b0}\u{95fb}", "\u{5929}\u{6c14}",
    "\u{8c03}\u{67e5}", "\u{641c}\u{7d22}", "\u{8054}\u{7f51}"
  ]
  private static let androidTerms = [
    "android", "apk", "mobile app", "phone game", "on the phone",
    "\u{5b89}\u{5353}", "\u{624b}\u{673a} app", "\u{624b}\u{673a}\u{4e0a}\u{73a9}",
    "\u{624b}\u{673a}\u{6e38}\u{620f}", "\u{5b89}\u{88c5}\u{5230}\u{624b}\u{673a}"
  ]
}

private extension String {
  func containsAny(_ terms: [String]) -> Bool {
    terms.contains { contains($0) }
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentExecutionPolicy.kt
package com.galaxyssi.chat

import java.util.Locale

enum class AgentExecutionTaskKind {
    CHAT,
    RESEARCH,
    ARTIFACT,
    BUILD,
    INSTALL,
    DEVICE
}

enum class AgentExecutionReasoningEffort {
    LOW,
    MEDIUM,
    HIGH
}

data class AgentExecutionProfile(
    val taskKind: AgentExecutionTaskKind,
    val reasoningEffort: AgentExecutionReasoningEffort,
    val noProgressTimeoutMillis: Long,
    val maxSameFailureAttempts: Int = 2,
    val requiresArtifact: Boolean = false,
    val targetPlatform: String = "",
    val verifyInstallation: Boolean = false,
    val taskIntent: AgentTaskIntent = AgentTaskIntent.CHAT,
    val taskIntentConfidence: Int = 100,
    val taskIntentSignals: List<String> = emptyList()
) {
    companion object {
        fun forGoal(goal: String, hasAttachments: Boolean = false): AgentExecutionProfile {
            val normalized = goal.lowercase(Locale.US).replace(Regex("\\s+"), " ").trim()
            val install = normalized.containsAny(INSTALL_TERMS)
            val build = normalized.containsAny(BUILD_TERMS)
            val artifactRequest = normalized.containsAny(ARTIFACT_TERMS)
            val research = normalized.containsAny(RESEARCH_TERMS)
            val device = normalized.containsAny(DEVICE_TERMS)
            val intent = AgentTaskIntentClassifier.classify(
                goal = normalized,
                hasAttachments = hasAttachments
            )
            val taskKind = when {
                install -> AgentExecutionTaskKind.INSTALL
                build -> AgentExecutionTaskKind.BUILD
                artifactRequest || hasAttachments -> AgentExecutionTaskKind.ARTIFACT
                research -> AgentExecutionTaskKind.RESEARCH
                device -> AgentExecutionTaskKind.DEVICE
                else -> AgentExecutionTaskKind.CHAT
            }
            val complex = taskKind in setOf(
                AgentExecutionTaskKind.RESEARCH,
                AgentExecutionTaskKind.ARTIFACT,
                AgentExecutionTaskKind.BUILD,
                AgentExecutionTaskKind.INSTALL
            )
            val timeoutMillis = when (taskKind) {
                AgentExecutionTaskKind.CHAT -> 180_000L
                AgentExecutionTaskKind.DEVICE -> 120_000L
                AgentExecutionTaskKind.RESEARCH -> 300_000L
                AgentExecutionTaskKind.ARTIFACT -> 360_000L
                AgentExecutionTaskKind.BUILD,
                AgentExecutionTaskKind.INSTALL -> 420_000L
            }
            return AgentExecutionProfile(
                taskKind = taskKind,
                reasoningEffort = if (complex) {
                    AgentExecutionReasoningEffort.MEDIUM
                } else {
                    AgentExecutionReasoningEffort.LOW
                },
                noProgressTimeoutMillis = timeoutMillis,
                requiresArtifact = artifactRequest || taskKind in setOf(
                    AgentExecutionTaskKind.BUILD,
                    AgentExecutionTaskKind.INSTALL
                ),
                targetPlatform = if (normalized.containsAny(ANDROID_TERMS)) "android" else "",
                verifyInstallation = taskKind == AgentExecutionTaskKind.INSTALL,
                taskIntent = intent.intent,
                taskIntentConfidence = intent.confidence,
                taskIntentSignals = intent.matchedSignals
            )
        }

        private val BUILD_TERMS = listOf(
            "build", "compile", "implement", "develop", "write a program", "create an app",
            "create a game", "fix bug", "run tests", "improve", "upgrade", "update", "refactor",
            "\u7f16\u8bd1", "\u6784\u5efa", "\u5f00\u53d1", "\u5b9e\u73b0",
            "\u5199\u4e00\u4e2a\u7a0b\u5e8f", "\u505a\u4e00\u4e2a\u6e38\u620f",
            "\u751f\u6210\u7a0b\u5e8f", "\u4fee\u590d bug", "\u8fd0\u884c\u6d4b\u8bd5",
            "\u6539\u8fdb", "\u5347\u7ea7", "\u66f4\u65b0", "\u91cd\u6784"
        )
        private val INSTALL_TERMS = listOf(
            "install", "install and open", "install apk", "deploy to phone", "launch the app",
            "\u5b89\u88c5", "\u5b89\u88c5\u5e76\u6253\u5f00", "\u5b89\u88c5 apk",
            "\u5b89\u88c5\u5230\u624b\u673a", "\u7f16\u8bd1\u5e76\u5b89\u88c5"
        )
        private val ARTIFACT_TERMS = listOf(
            "return the file", "send the file", "export", "generate image", "create file",
            "downloadable", "zip project", "apk",
            "\u53d1\u56de\u6587\u4ef6", "\u8fd4\u56de\u6587\u4ef6", "\u5bfc\u51fa",
            "\u751f\u6210\u56fe\u7247", "\u6253\u5305", "\u538b\u7f29\u5305"
        )
        private val RESEARCH_TERMS = listOf(
            "latest", "today", "news", "weather", "research", "search the web",
            "\u6700\u65b0", "\u4eca\u5929", "\u65b0\u95fb", "\u5929\u6c14",
            "\u8c03\u67e5", "\u641c\u7d22", "\u8054\u7f51"
        )
        private val DEVICE_TERMS = listOf(
            "battery", "flashlight", "camera", "alarm", "timer", "phone setting",
            "\u7535\u91cf", "\u624b\u7535\u7b52", "\u6444\u50cf\u5934", "\u62cd\u7167",
            "\u95f9\u949f", "\u8ba1\u65f6\u5668", "\u624b\u673a\u8bbe\u7f6e"
        )
        private val ANDROID_TERMS = listOf(
            "android", "apk", "mobile app", "phone game", "on the phone",
            "\u5b89\u5353", "\u624b\u673a app", "\u624b\u673a\u4e0a\u73a9",
            "\u624b\u673a\u6e38\u620f", "\u5b89\u88c5\u5230\u624b\u673a"
        )
    }
}

internal fun String.containsAny(terms: Iterable<String>): Boolean =
    terms.any(::contains)

internal fun AgentExecutionProfile.contract(): String = buildString {
    append("GalaxySSI execution contract: task=")
        .append(taskKind.name.lowercase(Locale.US))
        .append(", intent=")
        .append(taskIntent.name.lowercase(Locale.US))
        .append(", reasoning_effort=")
        .append(reasoningEffort.name.lowercase(Locale.US))
        .append(". Use Plan -> Act -> Observe -> Replan -> Verify -> Finalize. ")
    append("Checkpoint useful work before long or risky actions. ")
    append("Do not repeat an unchanged failing approach. ")
    if (requiresArtifact) {
        append("A single deliverable remains in its native format; package a directory or multi-file project as ZIP. ")
    }
    if (verifyInstallation) {
        append("Only report installation or launch after Android returns a verified execution receipt. ")
    }
    append("Do not report success without verification evidence.")
}
